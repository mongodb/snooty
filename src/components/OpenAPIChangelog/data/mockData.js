export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
