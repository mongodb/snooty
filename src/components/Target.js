import React from 'react';
import PropTypes from 'prop-types';
import ComponentFactory from './ComponentFactory';

// Based on condition isValid, split array into two arrays: [[valid, invalid]]
const partition = (array, isValid) => {
  return array.reduce(
    ([pass, fail], elem) => {
      return isValid(elem) ? [[...pass, elem], fail] : [pass, [...fail, elem]];
    },
    [[], []]
  );
};

const DescriptionTerm = ({ children, html_id, ...rest }) => {
  return (
    <dt id={html_id}>
      {children.map((child, j) => (
        <ComponentFactory key={j} {...rest} nodeData={child} />
      ))}
      <a href={`#${html_id}`} className="headerlink" title="Permalink to this definition">
        Â¶
      </a>
    </dt>
  );
};

DescriptionTerm.propTypes = {
  children: PropTypes.arrayOf(PropTypes.object).isRequired,
  html_id: PropTypes.string.isRequired,
};

const Target = ({ nodeData: { children, html_id, name }, ...rest }) => {
  // If directive_argument node is not present, render an empty span with the target ID
  // Otherwise, render directive_argument as a dictionary node and attach the
  // ID to the description term field
  const [, dictList] = partition(children, (elem) => elem.type === 'target_identifier');
  const [[descriptionTerm], descriptionDetails] = partition(dictList, (elem) => elem.type === 'directive_argument');

  return (
    <React.Fragment>
      {/* Render binary and program targets as empty spans such that their IDs are rendered on the page. */}
      {dictList.length > 0 && !['binary', 'program'].includes(name) ? (
        <dl className={name}>
          {descriptionTerm && <DescriptionTerm {...rest} {...descriptionTerm} html_id={html_id} />}
          <dd>
            {descriptionDetails.map((node, i) => (
              <ComponentFactory {...rest} nodeData={node} key={i} />
            ))}
          </dd>
        </dl>
      ) : (
        <span id={html_id} />
      )}
    </React.Fragment>
  );
};

Target.propTypes = {
  nodeData: PropTypes.shape({
    children: PropTypes.arrayOf(PropTypes.object).isRequired,
    html_id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
};

export default Target;
