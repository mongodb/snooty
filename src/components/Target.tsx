import React, { useRef } from 'react';
import { cx, css } from '@leafygreen-ui/emotion';
import useHashAnchor from '../hooks/use-hash-anchor';
import { theme } from '../theme/docsTheme';
import type { TargetNode, ASTNode } from '../types/ast';
import ComponentFactory from './ComponentFactory';
import Permalink from './Permalink';

const headerBuffer = css`
  scroll-margin-top: ${theme.header.navbarScrollOffset};
  position: absolute;
`;

// Based on condition isValid, split array into two arrays: [[valid, invalid]]
const partition = (array: ASTNode[], isValid: (elem: ASTNode) => boolean) => {
  return array.reduce(
    ([pass, fail], elem) => {
      return isValid(elem) ? [[...pass, elem], fail] : [pass, [...fail, elem]];
    },
    [[] as ASTNode[], [] as ASTNode[]]
  );
};

interface DescriptionTermProps {
  children?: ASTNode[];
  html_id: string;
}

const DescriptionTerm = ({ children, html_id, ...rest }: DescriptionTermProps) => {
  return (
    <dt>
      {children?.map((child, j) => (
        <ComponentFactory key={j} {...rest} nodeData={child} />
      ))}
      <Permalink id={html_id} description="definition" />
    </dt>
  );
};

interface TargetProps {
  nodeData: TargetNode;
}

const Target = ({ nodeData: { children, html_id, name, options }, ...rest }: TargetProps) => {
  // If directive_argument node is not present, render an empty span with the target ID
  // Otherwise, render directive_argument as a dictionary node and attach the
  // ID to the description term field
  const [, dictList] = partition(children, (elem) => elem.type === 'target_identifier');
  const [[descriptionTerm], descriptionDetails] = partition(dictList, (elem) => elem.type === 'directive_argument');
  const hidden = options && options.hidden ? true : false;
  const targetRef = useRef<HTMLSpanElement>(null);
  useHashAnchor(html_id, targetRef.current);

  return (
    <React.Fragment>
      {/* Render binary and program targets **and targets with the :hidden: flag
      as empty spans such that their IDs are rendered on the page. */}
      {dictList.length > 0 && !['binary', 'program'].includes(name) && !hidden ? (
        <dl className={name}>
          {descriptionTerm && <DescriptionTerm {...rest} {...descriptionTerm} html_id={html_id} />}
          <dd>
            {descriptionDetails.map((node, i) => (
              <ComponentFactory {...rest} nodeData={node} key={i} />
            ))}
          </dd>
        </dl>
      ) : (
        <span ref={targetRef} className={cx(headerBuffer)} id={html_id} />
      )}
    </React.Fragment>
  );
};

export default Target;
