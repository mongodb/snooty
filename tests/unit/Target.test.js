import React from 'react';
import { render } from 'enzyme';
import Target from '../../src/components/Target';

// data for this component
import mockData from './data/Target.test.json';

it('renders correctly with a directive_argument node', () => {
  const tree = render(<Target nodeData={mockData.a} />);
  expect(tree).toMatchSnapshot();
});

it('renders correctly with no directive_argument nodes', () => {
  const tree = render(<Target nodeData={mockData.b} />);
  expect(tree).toMatchSnapshot();
});
