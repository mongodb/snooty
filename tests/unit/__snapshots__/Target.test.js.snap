// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with a directive_argument node 1`] = `
.emotion-2 {
  background-color: #F9FBFA;
  border: 1px solid #B8C4C2;
  border-radius: 3px;
  font-family: 'Source Code Pro',Menlo,monospace;
  font-size: 14px;
  line-height: 20px;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  white-space: nowrap;
}

[data-leafygreen-ui="anchor-inline-code"]:hover > .emotion-2 {
  box-shadow: 0 0 0 3px #E7EEEC;
}

[data-leafygreen-ui="anchor-inline-code"]:focus > .emotion-2 {
  box-shadow: 0 0 0 3px #C5E4F2;
}

.emotion-0 {
  font-size: unset;
}

<dl
  class="option"
>
  <dt
    id="std-option-mongos.--config"
  >
    <code
      class="emotion-0 emotion-1 emotion-2"
    >
      --config &lt;filename&gt;, -f &lt;filename&gt;
    </code>
    <a
      class="headerlink"
      href="#std-option-mongos.--config"
      title="Permalink to this definition"
    >
      Â¶
    </a>
  </dt>
  <dd>
    <p>
      Specifies a configuration file for runtime configuration options. The
configuration file is the preferred method for runtime configuration of

      <a
        href="/reference/program/mongos/#std-program-mongos"
      >
        <code
          class="emotion-0 emotion-1 emotion-2"
        >
          mongos
        </code>
      </a>
      . The options are equivalent to the command-line
configuration options. See 
      <a
        href="/reference/configuration-options/"
      >
        Configuration File Options
      </a>
       for
more information.
    </p>
    <p>
      Ensure the configuration file uses ASCII encoding. The 
      <a
        href="/reference/program/mongos/#std-program-mongos"
      >
        <code
          class="emotion-0 emotion-1 emotion-2"
        >
          mongos
        </code>
      </a>
      
instance does not support configuration files with non-ASCII encoding,
including UTF-8.
    </p>
  </dd>
</dl>
`;

exports[`renders correctly with no directive_argument nodes 1`] = `
<span
  id="std-label-php-language-center"
/>
`;
